src/storage/v2/disk/storage.cpp

```diff
@@ -774,98 +774,278 @@ std::optional<VertexAccessor> DiskStorage::DiskAccessor::FindVertex(storage::Gid
-    auto ret = DeleteEdge(&e);
-    if (ret.HasError()) {
-      MG_ASSERT(ret.GetError() == Error::SERIALIZATION_ERROR, "Invalid database state!");
-      return ret.GetError();
+  std::set<EdgeRef> deleted_edges_set;
+
+  // routine for deleting edges, please use this under the edge spin lock
+  auto mark_edge_as_deleted_func = [this, &deleted_edges, &deleted_edges_set](auto edge_ref, auto edge_type,
+                                                                              auto *from_vertex, auto *to_vertex) {
+    if (config_.properties_on_edges) {
+      auto *edge_ptr = edge_ref.ptr;
+      if (!edge_ptr->deleted) {
+        CreateAndLinkDelta(&transaction_, edge_ptr, Delta::RecreateObjectTag());
+        edge_ptr->deleted = true;
+        storage_->edge_count_.fetch_add(-1, std::memory_order_acq_rel);
+      }
     }
 
-    if (ret.GetValue()) {
-      deleted_edges.push_back(*ret.GetValue());
+    // edges need to be deleted twice from the src vertex vector and the dest vertex vector, which
+    // guarantees that the deleted edges will be updated
+    if (!deleted_edges_set.insert(edge_ref).second) {
+      deleted_edges.emplace_back(edge_ref, edge_type, from_vertex, to_vertex, &transaction_, &storage_->indices_,
+                                 &storage_->constraints_, config_, true);
+
+      const std::string src_dest_del_key{
+          utils::SerializeEdge(from_vertex->gid, to_vertex->gid, edge_type, edge_ref, config_.properties_on_edges)};
+      edges_to_delete_.emplace_back(src_dest_del_key);
     }
+  };
+
+  // Gather nodes for deletion
+  std::set<Vertex *> nodes_to_delete;
+  {
+    for (const auto *vertex : nodes) {
+      MG_ASSERT(vertex->transaction_ == &transaction_,
+                "VertexAccessor must be from the same transaction as the storage "
+                "accessor when deleting a vertex!");
+      auto *vertex_ptr = vertex->vertex_;
+
+      {
+        std::lock_guard<utils::SpinLock> guard(vertex_ptr->lock);
+
+        if (!PrepareForWrite(&transaction_, vertex_ptr)) return Error::SERIALIZATION_ERROR;
+
+        if (vertex_ptr->deleted) {
+          continue;
+        }
+      }
+
+      nodes_to_delete.insert(vertex_ptr);
+    }
+
+    deleted_vertices.reserve(nodes_to_delete.size());
   }
-  for (const auto &item : out_edges) {
-    auto [edge_type, to_vertex, edge] = item;
-    EdgeAccessor e(edge, edge_type, vertex_ptr, to_vertex, &transaction_, &storage_->indices_, &storage_->constraints_,
-                   config_);
-    auto ret = DeleteEdge(&e);
-    if (ret.HasError()) {
-      MG_ASSERT(ret.GetError() == Error::SERIALIZATION_ERROR, "Invalid database state!");
-      return ret.GetError();
+
+  // Aggregate edges and nodes for partial deletion (edges of nodes which don't need deletion)
+  std::set<Vertex *> partial_src_vertices;
+  std::set<Vertex *> partial_dest_vertices;
+  std::set<EdgeRef> src_edge_refs;
+  std::set<EdgeRef> dest_edge_refs;
+  auto total_edges_to_delete = 0;
```

Comment:
I know you like auto and everything but let's try to reduce this, it's super hard to read so many autos in arguments